### PR TITLE
Sidebar Tabs: Use editor settings to override display

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:
- https://github.com/WordPress/gutenberg/pull/45005

## What?

- Leverages a new block editor setting to provide a means of overriding the block inspector tabs display across all blocks or on a per-block basis.
- Uses the new block editor setting to disable tabs for the Navigation Link block

## Why?

We need a means of providing plugins and third party blocks with the ability to disable the display of block inspector tabs.

## How?

- Adds a new experimental block editor setting: `__experimentalBlockInspectorTabs`
- Adds a `block_editor_settings_all` filter to the Navigation Link block to disable tabs for it
- Updates the block inspector to take into account any overrides in the new block editor setting when rendering tabs
- Updates the item schema for the block editor settings rest controller

#### Alternative Approaches
- Using a filter: https://github.com/WordPress/gutenberg/pull/46026 
- Using the block editor store: https://github.com/WordPress/gutenberg/pull/46271
- A context approach was considered however we need to be able to set the overrides outside of components within the context.

## Testing Instructions
1. Enable both the "off canvas navigation editor" and "block inspector tabs" Gutenberg experiments
2. Edit a post
3. Add a navigation block and a navigation link
4. Ensure the navigation block is selected
5. Under the navigation block's list view tab edit the link
6. You should now have the navigation link block selected without any display of tabs in the block inspector

### Testing Instructions for Keyboard
1. Enable both the "off canvas navigation editor" and "block inspector tabs" Gutenberg experiments
2. Edit a post
3. Add a navigation block and a navigation link
4. Ensure the navigation block is selected
5. Tab through to the navigation block's list view tab and focus the navigation link's edit button
6. Press space and you should then have the navigation link block selected with no tabs displayed in the block inspector

## Screenshots or screencast 

https://user-images.githubusercontent.com/60436221/205834991-c670cce5-c03e-4202-a24e-6d9fb1fcd358.mp4

### CODE REVIEW

File: packages/block-editor/src/components/block-inspector/index.js [hunk 1]
```diff
@@ -72,6 +72,18 @@ function getContentBlocks( blocks, isContentBlock ) {
...
+function getShowTabs( blockName, tabSettings = {} ) {
+	if ( tabSettings[ blockName ] !== undefined ) {
+		return tabSettings[ blockName ];
+	}
+
+	if ( tabSettings.default !== undefined ) {
+		return tabSettings.default;
+	}
+
+	return window?.__experimentalEnableBlockInspectorTabs;
```
Comment (talldan) [inline packages/block-editor/src/components/block-inspector/index.js]: I wonder if this should be the first thing checked in the function? At the moment the setting can bypass the experiment. The experiment should maybe overrule everything else.

File: packages/block-editor/src/components/block-inspector/index.js [hunk 2]
```diff
@@ -242,10 +257,14 @@ const BlockInspector = ( { showNoBlockSelectedMessage = true } ) => {
 };
 
 const BlockInspectorSingleBlock = ( { clientId, blockName } ) => {
+	const tabSettings = useSelect( ( select ) => {
+		return select( blockEditorStore ).getSettings()
+			.__experimentalBlockInspectorTabs;
+	}, [] );
```
Comment (talldan) [inline packages/block-editor/src/components/block-inspector/index.js]: It might be possible for this and `getShowTabs` to be folded into the `useInspectorControlsTabs` hook.

File: packages/block-editor/src/components/block-inspector/index.js [hunk 3]
```diff
@@ -72,6 +72,18 @@ function getContentBlocks( blocks, isContentBlock ) {
...
 }
 
+function getShowTabs( blockName, tabSettings = {} ) {
+	if ( tabSettings[ blockName ] !== undefined ) {
+		return tabSettings[ blockName ];
+	}
+
+	if ( tabSettings.default !== undefined ) {
+		return tabSettings.default;
+	}
```
Comment (talldan) [inline packages/block-editor/src/components/block-inspector/index.js]: So at the moment you can set `default: false` and something like `'core/paragraph': true` to enable tabs only on one block. Or `default: true` and `'core/paragraph': false` to disable only one block.

That's pretty flexible, and honors the individual tab setting, which is nice.

The other option might be having 'default' as a separate setting entirely instead of a property, but I don't really have a strong preference either way. I wonder if anyone has ever made a block with the name 'default', as that may cause an issue. It would be going against the block naming guidelines, which recommend namespaced block names (my-plugin/default), so probably not something to worry about.

File: packages/block-library/src/navigation-link/index.php [hunk 4]
```diff
@@ -344,3 +344,26 @@ function register_block_core_navigation_link() {
...
+function gutenberg_disable_tabs_for_navigation_link_block( $settings ) {
+	$current_tab_settings = _wp_array_get(
+		$settings,
+		array( '__experimentalBlockInspectorTabs' ),
+		array()
+	);
+
+	$settings['__experimentalBlockInspectorTabs'] = array_merge(
+		$current_tab_settings,
+		array( 'core/navigation-link' => false )
```
Comment (talldan) [inline packages/block-library/src/navigation-link/index.php:363]: One thing is that this should probably also be applied to `core/navigation-submenu`, the other main block type in the navigation list view. (cc @getdave @scruffian @draganescu @MaggieCabrera for thoughts on this)

Comment (talldan) [reply packages/block-library/src/navigation-link/index.php:363]: Actually, just testing and I notice submenu already doesn't have any tabs (it has no appearance tools).

It might still be worth disabling tabs though, as that's not a reliable situation (plugins can modify things, and the situation might change in the future).

**Review by talldan — COMMENTED** [review]
Pretty happy with this approach.

I personally don't think removing tabs from the link block is the long term solution to the nav block's List View interface, but it does solve the immediate problem and provides some time for more exploration into the right solution.

Nice work!

File: packages/block-library/src/navigation-link/index.php (hunk 4 above)
Comment (scruffian) [reply packages/block-library/src/navigation-link/index.php:363]: +1 to disabling on navigation-submenu

File: packages/block-editor/src/components/block-inspector/index.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: ✅ Done. https://github.com/WordPress/gutenberg/pull/46321/commits/e5dd52f58bbf8de31645fbff93d4f85aedbf56d0

The editor settings now have no effect unless the Gutenberg experiment for block inspector tabs has been opted into.

File: packages/block-editor/src/components/block-inspector/index.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: Thanks. Nice to have the tab stuff in the one place as much as possible 👍 

✅ Done: https://github.com/WordPress/gutenberg/pull/46321/commits/e5dd52f58bbf8de31645fbff93d4f85aedbf56d0

File: packages/block-editor/src/components/block-inspector/index.js (hunk 3 above)
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: Thanks. I'll be happy to trial the alternative if others have stronger opinions or different thoughts.

File: packages/block-library/src/navigation-link/index.php (hunk 4 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/navigation-link/index.php:363]: ✅ Done: https://github.com/WordPress/gutenberg/pull/46321/commits/7d4e31fa9964a1445385c729b0563945342b259a

File: packages/block-editor/src/components/block-inspector/index.js [hunk 5]
```diff
@@ -169,9 +169,7 @@ const BlockInspector = ( { showNoBlockSelectedMessage = true } ) => {
 	}, [] );
 
 	const availableTabs = useInspectorControlsTabs( blockType?.name );
-	const showTabs =
-		window?.__experimentalEnableBlockInspectorTabs &&
-		availableTabs.length > 1;
+	const showTabs = availableTabs.length > 1;
```
Comment (getdave) [inline packages/block-editor/src/components/block-inspector/index.js:172]: Will `availableTabs` always have a `length` property?

File: packages/block-editor/src/components/block-inspector/index.js [hunk 6]
```diff
@@ -243,9 +241,7 @@ const BlockInspector = ( { showNoBlockSelectedMessage = true } ) => {
 
 const BlockInspectorSingleBlock = ( { clientId, blockName } ) => {
 	const availableTabs = useInspectorControlsTabs( blockName );
-	const showTabs =
-		window?.__experimentalEnableBlockInspectorTabs &&
-		availableTabs.length > 1;
+	const showTabs = availableTabs.length > 1;
```
Comment (getdave) [inline packages/block-editor/src/components/block-inspector/index.js:244]: Will `availableTabs` always have a `length` property?

File: packages/block-editor/src/components/inspector-controls-tabs/use-inspector-controls-tabs.js [hunk 7]
```diff
@@ -54,5 +76,12 @@ export default function useInspectorControlsTabs( blockName ) {
...
 
-	return tabs;
+	const tabSettings = useSelect( ( select ) => {
+		return select( blockEditorStore ).getSettings()
+			.__experimentalBlockInspectorTabs;
+	}, [] );
+
+	const showTabs = getShowTabs( blockName, tabSettings );
+
+	return showTabs ? tabs : [];
```
Comment (getdave) [inline packages/block-editor/src/components/inspector-controls-tabs/use-inspector-controls-tabs.js:86]: Shall we return a const `EMPTY_ARRAY` from outside the hook scope? That way we have a stable reference for the hook to avoid unnecessary changes.

File: packages/block-library/src/navigation-link/index.php [hunk 8]
```diff
@@ -344,3 +344,26 @@ function register_block_core_navigation_link() {
 	);
 }
 add_action( 'init', 'register_block_core_navigation_link' );
+
+/**
+ * Disables the display of tabs for the Navigation Link block.
```
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php:349]: Can we add some context as to why this decision has been taken?

Let's add a link to the discussions that landed us here. I would make a suggestion but I've lost track of where...

**Review by getdave — COMMENTED** [review]
Thank you so much for working on this.

Some things I spotted. Will test manually shortly...

**Review by getdave — APPROVED** [review]
https://user-images.githubusercontent.com/444434/205897515-5d15fcbd-1ef0-4a9f-83cc-84eef6e1eb16.mp4

I tested in both Site and Post editors and this worked as expected for the Nav block experiment 👏

File: packages/block-library/src/navigation-submenu/index.php [hunk 9]
```diff
@@ -289,3 +289,26 @@ function register_block_core_navigation_submenu() {
...
 }
 add_action( 'init', 'register_block_core_navigation_submenu' );
+
+/**
+ * Disables the display of tabs for the Navigation Submenu block.
+ *
+ * @param array $settings Default editor settings.
+ * @return array Filtered editor settings.
+ */
+function gutenberg_disable_tabs_for_navigation_submenu_block( $settings ) {
```
Comment (scruffian) [inline packages/block-library/src/navigation-submenu/index.php:299]: is it worth having a shared file for functions that both blocks use?

**Review by scruffian — APPROVED** [review]
LGTM

**Comment by scruffian** [comment]
Lets address the feedback in a followup...

File: packages/block-library/src/navigation-submenu/index.php (hunk 9 above)
Comment (getdave) [reply packages/block-library/src/navigation-submenu/index.php:299]: TBH these blocks are so similar it would also make sense to share JS code as well. Is there a precedent for this? 

Can we make one block canonical and import the code into the other block from the other's directory?

Comment (aaronrobertshaw) [reply packages/block-library/src/navigation-submenu/index.php:299]: > is it worth having a shared file for functions that both blocks use?

I did look at something along these lines in the example diff below. Ultimately, it doesn't seem worth the change, the potential clutter in core etc. Merging the two blocks almost into a block variation type situation sounds like a better option down the road.



<details>
<summary>Example Diff</summary>

```diff
diff --git a/lib/experimental/editor-settings.php b/lib/experimental/editor-settings.php
index 862a1bfe2b..d36dce2d14 100644
--- a/lib/experimental/editor-settings.php
+++ b/lib/experimental/editor-settings.php
@@ -72,6 +72,29 @@ function gutenberg_initialize_editor( $editor_name, $editor_script_handle, $sett
 	);
 }
 
+/**
+ * Flags the specified block to disable block inspector tabs.
+ *
+ * @param string $block_name Block name.
+ * @param array  $settings   Default editor settings.
+ *
+ * @return array Filtered editor settings.
+ */
+function gutenberg_disable_block_inspector_tabs_for_block( $block_name, $settings ) {
+	$current_tab_settings = _wp_array_get(
+		$settings,
+		array( '__experimentalBlockInspectorTabs' ),
+		array()
+	);
+
+	$settings['__experimentalBlockInspectorTabs'] = array_merge(
+		$current_tab_settings,
+		array( $block_name => false )
+	);
+
+	return $settings;
+}
+
 /**
  * Sets a global JS variable used to trigger the availability of each Gutenberg Experiment.
  */
diff --git a/packages/block-library/src/navigation-link/index.php b/packages/block-library/src/navigation-link/index.php
index 8853d49f6e..a558c49e83 100644
--- a/packages/block-library/src/navigation-link/index.php
+++ b/packages/block-library/src/navigation-link/index.php
@@ -361,18 +361,7 @@ add_action( 'init', 'register_block_core_navigation_link' );
  * @return array Filtered editor settings.
  */
 function gutenberg_disable_tabs_for_navigation_link_block( $settings ) {
-	$current_tab_settings = _wp_array_get(
-		$settings,
-		array( '__experimentalBlockInspectorTabs' ),
-		array()
-	);
-
-	$settings['__experimentalBlockInspectorTabs'] = array_merge(
-		$current_tab_settings,
-		array( 'core/navigation-link' => false )
-	);
-
-	return $settings;
+	return gutenberg_disable_block_inspector_tabs_for_block( 'core/navigation-link', $settings );
 }
 
 add_filter( 'block_editor_settings_all', 'gutenberg_disable_tabs_for_navigation_link_block' );
diff --git a/packages/block-library/src/navigation-submenu/index.php b/packages/block-library/src/navigation-submenu/index.php
index 99c86493d5..25de8de0e3 100644
--- a/packages/block-library/src/navigation-submenu/index.php
+++ b/packages/block-library/src/navigation-submenu/index.php
@@ -306,18 +306,7 @@ add_action( 'init', 'register_block_core_navigation_submenu' );
  * @return array Filtered editor settings.
  */
 function gutenberg_disable_tabs_for_navigation_submenu_block( $settings ) {
-	$current_tab_settings = _wp_array_get(
-		$settings,
-		array( '__experimentalBlockInspectorTabs' ),
-		array()
-	);
-
-	$settings['__experimentalBlockInspectorTabs'] = array_merge(
-		$current_tab_settings,
-		array( 'core/navigation-submenu' => false )
-	);
-
-	return $settings;
+	return gutenberg_disable_block_inspector_tabs_for_block( 'core/navigation-submenu', $settings );
 }
 
 add_filter( 'block_editor_settings_all', 'gutenberg_disable_tabs_for_navigation_submenu_block' );

```
</details>

**Comment by aaronrobertshaw** [comment]
A PR addressing the majority of feedback is up in https://github.com/WordPress/gutenberg/pull/46346
